File: CrowdAssist/app/src/main/java/th/ac/kmitl/it/crowdalert/RateActivity.java
package th.ac.kmitl.it.crowdalert;

import android.arch.lifecycle.LiveData;
import android.arch.lifecycle.Observer;
import android.arch.lifecycle.ViewModelProviders;
import android.content.Context;
import android.content.SharedPreferences;
import android.net.Uri;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v7.app.AppCompatActivity;
import android.util.Log;
import android.view.View;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.RatingBar;
import android.widget.TextView;

import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.OnSuccessListener;
import com.google.android.gms.tasks.Task;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.ValueEventListener;
import com.google.firebase.storage.FirebaseStorage;
import com.squareup.picasso.Picasso;

import java.util.LinkedList;
import java.util.Queue;

import th.ac.kmitl.it.crowdalert.model.AssistantModel;
import th.ac.kmitl.it.crowdalert.model.UserModel;
import th.ac.kmitl.it.crowdalert.util.AssistantViewModel;
import th.ac.kmitl.it.crowdalert.util.DatabaseHelper;

public class RateActivity extends AppCompatActivity implements View.OnClickListener, OnCompleteListener<Void> {
    private final String SP_REQUEST = "request_information";
    private FirebaseDatabase mDatabase;
    private FirebaseStorage mStorage;
    private SharedPreferences requestSP;
    private String requestUid;
    private String requestType;
    private String requesterUid;
    private AssistantModel currentUser;
    private DatabaseHelper helper;
    private Queue<AssistantModel> userQueue;
    private FirebaseUser mUser;
    private FirebaseAuth mAuth;
    private LinearLayout progressLayout;
    private LinearLayout rateCardLayout;
    private ImageView profileImage;
    private RatingBar ratingBar;
    private TextView nameTextView;
    private TextView roleTextView;
    private EditText descriptionEditText;
    private LinearLayout rateButton;
    private LinearLayout ignoreButton;
    private UserModel userModel;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_rate);
        helper = new DatabaseHelper(this);
        mAuth = FirebaseAuth.getInstance();
        mUser = mAuth.getCurrentUser();
        mDatabase = FirebaseDatabase.getInstance();
        mStorage = FirebaseStorage.getInstance();
        bindView();
        userQueue = new LinkedList<>();
        requestSP = getSharedPreferences(SP_REQUEST, Context.MODE_PRIVATE);

        requestUid = requestSP.getString("request_uid", null);
        requestType = requestSP.getString("type", null);
        requesterUid = requestSP.getString("requester_uid", null);

        progressLayout.setVisibility(View.VISIBLE);
        rateCardLayout.setVisibility(View.GONE);
        Log.d("Hello", mUser.getUid()+":"+requesterUid);
        if (mUser.getUid().equals(requesterUid)){
            AssistantViewModel viewModel = ViewModelProviders.of(this).get(AssistantViewModel.class);
            final LiveData<DataSnapshot> liveData = viewModel.getDataSnapshotLiveData();
            liveData.observe(this, new Observer<DataSnapshot>() {
                @Override
                public void onChanged(@Nullable DataSnapshot dataSnapshot) {
                    if (dataSnapshot == null){
                        Log.e("Assistant Livedata", "datasnapshot is empty");
                        setResult(RESULT_OK);
                        finish();
                    }
                    if (!dataSnapshot.exists()){
                        setResult(RESULT_OK);
                        finish();
                    }
                    if (dataSnapshot.exists()){
                        for (DataSnapshot data : dataSnapshot.getChildren()){
                            AssistantModel model = new AssistantModel();
                            model.setAssistantUid(data.getKey());
                            model.setRole(data.child("role").getValue(String.class));
                            push(model);
                        }
                    }
                    if (userQueue.isEmpty()) {
                        setResult(RESULT_OK);
                        finish();
                    }else{
                        currentUser = userQueue.remove();
                        liveData.removeObservers(RateActivity.this);
                        prepareCardView();
                    }
                }
            });
        }else{
            AssistantModel model = new AssistantModel();
            model.setAssistantUid(requesterUid);
            model.setRole("requester");
            push(model);
            currentUser = userQueue.remove();
            prepareCardView();
        }
    }

    @Override
    public void onClick(View view) {
        switch(view.getId()){
            case R.id.rateButton:
                progressLayout.setVisibility(View.VISIBLE);
                rateCardLayout.setVisibility(View.GONE);
                helper.rate(requestUid, currentUser.getAssistantUid(), Double.parseDouble(Float.toString(ratingBar.getRating())), descriptionEditText.getText().toString(), this);
                break;
            case R.id.ignoreButton:
                progressLayout.setVisibility(View.VISIBLE);
                rateCardLayout.setVisibility(View.GONE);
                profileImage.setImageResource(R.drawable.bg_gray);
                if (userQueue.isEmpty()) {
                    setResult(RESULT_OK);
                    finish();
                }else {
                    currentUser = userQueue.remove();
                    prepareCardView();
                }
                break;
        }
        //forDebug();
    }

    private void bindView(){
        progressLayout = findViewById(R.id.progress_layout);
        rateCardLayout = findViewById(R.id.rateCardLayout);
        profileImage = findViewById(R.id.imageView);
        nameTextView = findViewById(R.id.name);
        roleTextView = findViewById(R.id.role);
        ratingBar = findViewById(R.id.ratingBar);
        descriptionEditText = findViewById(R.id.description_text);
        rateButton = findViewById(R.id.rateButton);
        ignoreButton = findViewById(R.id.ignoreButton);
        rateButton.setOnClickListener(this);
        ignoreButton.setOnClickListener(this);
    }

    private void prepareCardView(){
        mDatabase.getReference().child("users").child(currentUser.getAssistantUid()).addListenerForSingleValueEvent(new ValueEventListener() {
            @Override
            public void onDataChange(DataSnapshot dataSnapshot) {
                if (dataSnapshot.exists()){
                    userModel = dataSnapshot.getValue(UserModel.class);
                    nameTextView.setText(String.format("%s %s", userModel.getFirstName(), userModel.getLastName()));
                    if ("requester".equals(currentUser.getRole())){
                        roleTextView.setText("ผู้ร้องขอ");
                    }else{
                        roleTextView.setText(String.format("%s", currentUser.getRole().equals("user") ? "อาสาสมัคร":"เข้าหน้าที่"));
                    }
                    mStorage.getReference().child("profile").child(currentUser.getAssistantUid()+".jpg").getDownloadUrl().addOnSuccessListener(new OnSuccessListener<Uri>() {
                        @Override
                        public void onSuccess(Uri uri) {
                            Picasso.with(RateActivity.this)
                                    .load(uri)
                                    .resize(500, 500)
                                    .into(profileImage);
                            descriptionEditText.setText("");
                            ratingBar.setRating(3);
                            progressLayout.setVisibility(View.GONE);
                            rateCardLayout.setVisibility(View.VISIBLE);
                        }
                    });
                }
            }

            @Override
            public void onCancelled(DatabaseError databaseError) {

            }
        });
    }

    private void push(AssistantModel val){
        userQueue.add(val);
    }

    @Override
    public void onComplete(@NonNull Task<Void> task) {
        if (task.isSuccessful()){
            if (userQueue.isEmpty()) {
                setResult(RESULT_OK);
                finish();
            }else {
                profileImage.setImageResource(R.drawable.bg_gray);
                currentUser = userQueue.remove();
                prepareCardView();
            }
        }
    }

    private void forDebug(){
        if (userQueue.isEmpty()) {
            finish();
        }else {
            profileImage.setImageResource(R.drawable.bg_gray);
            currentUser = userQueue.remove();
            prepareCardView();
        }
    }
}

File: CrowdAssist/app/src/main/java/th/ac/kmitl/it/crowdassist/RateActivity.kt
package th.ac.kmitl.it.crowdassist

import android.app.Activity
import android.arch.lifecycle.Observer
import android.arch.lifecycle.ViewModelProviders
import android.content.Context
import android.content.SharedPreferences
import android.net.Uri
import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.view.View
import android.widget.*
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.auth.FirebaseUser
import com.google.firebase.database.DataSnapshot
import com.squareup.picasso.Picasso
import th.ac.kmitl.it.crowdassist.contract.RateContract
import th.ac.kmitl.it.crowdassist.model.AssistantModel
import th.ac.kmitl.it.crowdassist.presenter.RatePresenter
import th.ac.kmitl.it.crowdassist.util.AssistantViewModel
import th.ac.kmitl.it.crowdassist.util.DatabaseHelper
import java.util.*

class RateActivity : AppCompatActivity(), RateContract.View{
    private val SP_REQUEST = "request_information"
    private var requestSP: SharedPreferences? = null
    private var requestUid: String? = null
    private var requestType: String? = null
    private var requesterUid: String? = null
    private var userQueue: Queue<AssistantModel>? = null
    private var mUser: FirebaseUser? = null
    private var mAuth: FirebaseAuth? = null
    private var progressLayout: LinearLayout? = null
    private var rateCardLayout: LinearLayout? = null
    private var profileImage: ImageView? = null
    private var ratingBar: RatingBar? = null
    private var nameTextView: TextView? = null
    private var roleTextView: TextView? = null
    private var descriptionEditText: EditText? = null
    private var rateButton: LinearLayout? = null
    private var ignoreButton: LinearLayout? = null
    private var presenter: RatePresenter? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_rate)
        requestSP = getSharedPreferences(SP_REQUEST, Context.MODE_PRIVATE)
        mAuth = FirebaseAuth.getInstance()
        mUser = mAuth?.currentUser

        requestUid = requestSP?.getString("request_uid", null)
        requestType = requestSP?.getString("type", null)
        requesterUid = requestSP?.getString("requester_uid", null)

        bindView()
        userQueue = LinkedList()
        showProgressBar()
        hideCardView()
        if (mUser?.uid == requesterUid) {
            val viewModel = ViewModelProviders.of(this).get(AssistantViewModel::class.java)
            val liveData = viewModel.dataSnapshotLiveData
            liveData.observe(this, Observer<DataSnapshot>{
                data: DataSnapshot? ->  kotlin.run {
                    data?.let {
                        if (!data.exists()){
                            finish(Activity.RESULT_OK)
                        }
                        if (data.exists()) {
                            for (item in data.children) {
                                val model = AssistantModel().also {
                                    it.assistantUid = item.key!!
                                    it.role = item.child("role").getValue(String::class.java)!!
                                }
                                push(model)
                            }
                        }
                        if (userQueue!!.isEmpty()) {
                            finish(Activity.RESULT_OK)
                        } else {
                            presenter = RatePresenter(DatabaseHelper(this), this, userQueue, requestUid!!)
                            presenter!!.prepareAssistantData()
                        }
                    } ?:run {
                        finish(Activity.RESULT_OK)
                    }
                }
            })
        }else{
            val model = AssistantModel().also {
                it.assistantUid = requesterUid!!
                it.role = "requester"
            }
            push(model)
            presenter!!.prepareAssistantData()
        }
    }

    private fun bindView() {
        progressLayout = findViewById(R.id.progress_layout)
        rateCardLayout = findViewById(R.id.rateCardLayout)
        profileImage = findViewById(R.id.imageView)
        nameTextView = findViewById(R.id.name)
        roleTextView = findViewById(R.id.role)
        ratingBar = findViewById(R.id.ratingBar)
        descriptionEditText = findViewById(R.id.description_text)
        rateButton = findViewById(R.id.rateButton)
        ignoreButton = findViewById(R.id.ignoreButton)
        rateButton!!.setOnClickListener{
            presenter?.onRateClicked()
        }
        ignoreButton!!.setOnClickListener{
            presenter?.onIgnoreClicked()
        }
    }

    override fun finish(resultCode: Int) {
        setResult(resultCode)
        finish()
    }

    override fun setNameCardView(name: String) {
        nameTextView?.text = name
    }

    override fun setRoleTextView(type: String) {
        roleTextView?.text = type
    }

    override fun setImageProfile(uri: Uri) {
        Picasso.with(this)
                .load(uri)
                .resize(500, 500)
                .into(profileImage)
    }

    override fun getRating(): Double {
        return ratingBar?.rating!!.toDouble()
    }

    override fun getDescription(): String {
        return descriptionEditText?.text.toString()
    }

    override fun showCardView() {
        rateCardLayout?.visibility = (View.VISIBLE)
    }

    override fun hideCardView() {
        rateCardLayout?.visibility = (View.GONE)
    }

    override fun showProgressBar() {
        progressLayout?.visibility = (View.VISIBLE)
    }

    override fun hideProgressBar() {
        progressLayout?.visibility = (View.GONE)
    }

    override fun setRating(rating: Double) {
        ratingBar?.rating = rating.toFloat()
    }

    override fun setDescription(description: String) {
        descriptionEditText?.setText(description)
    }

    private fun push(value: AssistantModel) {
        userQueue!!.add(value)
    }

    override fun setImageProfile(resId: Int) {
        profileImage?.setImageResource(resId)
    }
}
